File: src/cell.rs

use crate::util::PropagatorID;
use crate::content::Merge;
use std::fmt::Debug;

#[derive(Default, Debug)]
pub struct Cell<T> {
    pub label: String,
    content: T,
    neighbours: Vec<PropagatorID>,
}

impl<T: Default> Cell<T> {
    pub fn new() -> Self {
        Self::default()
    }
}

impl<T> Cell<T> {
    pub fn add_neighbour(&mut self, id: PropagatorID) {
        self.neighbours.push(id);
        self.neighbours.sort_unstable();
        self.neighbours.dedup();
    }

    pub fn read(&self) -> &T {
        &self.content
    }
}

impl<T: Merge + PartialEq + Debug> Cell<T> {
    pub fn write(&mut self, content: T) -> Option<&Vec<PropagatorID>> {
        let new_content = self.content.merge(&content);

        if self.content != new_content {
            self.content = new_content;

            Some(&self.neighbours)
        }
        else {
            None
        }
    }
}

File: src/cell.zig
const std = @import("std");
const warn = @import("std").debug.warn;
const PropagatorID = @import("util.zig").PropagatorID;
const ArrayList = std.ArrayList;
const Allocator = std.mem.Allocator;

pub fn Cell(comptime T: type) type {
    return struct {
        const Self = @This();

        name: []const u8,
        content: T,
        neighbours: ArrayList(PropagatorID),

        pub fn init(allocator: *Allocator) Self {
            return Self {
              .name = "cell", 
              .content = T.init(),
              .neighbours = ArrayList(PropagatorID).init(allocator)
            };
        }

        pub fn write(self: *Self, content: T) ?ArrayList(PropagatorID) {
            var new_content = self.content.merge(content);

            if (!self.content.equals(new_content)) {
              self.content = new_content;

              return self.neighbours;
            }
            else {
              return null;
            }
        }
    };
}
